test/unit/BPool/BPool_JoinPool.t.sol

```diff
@@ -0,0 +1,120 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.25;
+
+import {BPoolBase} from './BPoolBase.sol';
+import {IERC20} from '@openzeppelin/contracts/token/ERC20/IERC20.sol';
+
+import {BNum} from 'contracts/BNum.sol';
+import {IBPool} from 'interfaces/IBPool.sol';
+
+contract BPoolJoinPool is BPoolBase {
+  uint256 public poolAmountOut = 10e18;
+  // Assets Under Management
+  uint256 public tokenAUM = 10e18;
+  uint256 public secondTokenAUM = 30e18;
+
+  uint256 public requiredTokenIn = 1e18;
+  uint256 public requiredSecondTokenIn = 3e18;
+
+  function setUp() public virtual override {
+    super.setUp();
+    bPool.set__finalized(true);
+    // simulate expected finalize outcome
+    bPool.call__mintPoolShare(INIT_POOL_SUPPLY);
+    bPool.call__pushPoolShare(deployer, INIT_POOL_SUPPLY);
+    address[] memory _tokens = new address[](2);
+    _tokens[0] = token;
+    _tokens[1] = secondToken;
+    bPool.set__tokens(_tokens);
+    // token weights are not used for all-token joins
+    _setRecord(token, IBPool.Record({bound: true, index: 0, denorm: 0}));
+    _setRecord(secondToken, IBPool.Record({bound: true, index: 1, denorm: 0}));
+    // underlying balances are used instead
+    vm.mockCall(token, abi.encodePacked(IERC20.balanceOf.selector), abi.encode(uint256(tokenAUM)));
+    vm.mockCall(secondToken, abi.encodePacked(IERC20.balanceOf.selector), abi.encode(uint256(secondTokenAUM)));
+  }
+
+  function test_RevertWhen_ReentrancyLockIsSet() external {
+    bPool.call__setLock(_MUTEX_TAKEN);
+    // it should revert
+    vm.expectRevert(IBPool.BPool_Reentrancy.selector);
+    bPool.joinPool(0, new uint256[](2));
+  }
+
+  function test_RevertWhen_PoolIsNotFinalized() external {
+    bPool.set__finalized(false);
+    // it should revert
+    vm.expectRevert(IBPool.BPool_PoolNotFinalized.selector);
+    bPool.joinPool(0, new uint256[](2));
+  }
+
+  // should not happen in the real world since finalization mints 100 tokens
+  // and sends them to controller
+  function test_RevertWhen_TotalSupplyIsZero() external {
+    // undo what was just done by setup
+    bPool.call__pullPoolShare(deployer, INIT_POOL_SUPPLY);
+    bPool.call__burnPoolShare(INIT_POOL_SUPPLY);
```

Comment:
then why doing it on setup? 🤔

Reply:
because otherwise I'd have a deeper-branching tree, with most tests under `when pool has some initial supply` and, If I were consistent with setup-ing like that, `when pool is finalized`.

While doing that would be consistent code-wise, those cases are best treated like shallow branches, since those preconditions should be fullfilled by `bind` and `finalize`. Having them like this instead of a core precondition allows the tree to highlight the more relevant test cases & branches.

This is similar to how I set the pool as finalized in `setup`, since that's common to most cases, and set it as non-finalized in the body of `test_RevertWhen_PoolIsNotFinalized`